File: ECMWF/ERA5_snow.m
%% ECMWF datafile access
% example for access of data in a netCDF file and interpolation of a single coordinate.
% t2m: air temperature in 2m height
% sd: snow depth expressed as a water layer covering the grid cell
% sf: snowfall expressed as a water layer
clearvars
% Leipzig
pos=[51.3 12.37];
% plot all months
plotmonths=false;
%% Datafile downloaded with ERA5_snow.py
dfile='ERA5_snow_monthly025.nc';
ncdisp(dfile)
% Latitude and longitude from the datafile
lat=ncread(dfile,'latitude');
lon=ncread(dfile,'longitude');
% grid for plots withe geoshow
[glat,glon]=meshgrid(lat,lon);
% time from datafile converted to MATLAB datenum format
hzp=ncread(dfile,'time');
hzp=cast(hzp,'double')/24+datenum('1900/01/01 00:00:00');
%% Interpolation 
if plotmonths
    hh=figure('color',[1 1 1 ],'position',[100 100 600 1200]);
    wm=worldmap(double([min(lat) max(lat)]),double([min(lon) max(lon)]));
    cb=colorbar;
    caxis([0 13]);
    set(get(cb,'Label'),'string','Snow fall / mm water equiv.')
end
data=zeros(length(hzp),4);
expver=1; % new dimension in ERA5 Data 
% see https://confluence.ecmwf.int/pages/viewpage.action?pageId=173385064
for hi=1:length(hzp)
    d1=ncread(dfile,'t2m',[1 1 expver hi],[inf inf 1 1])-278.5651;
    if isnan(d1(1,1))
        expver=2;
        d1=ncread(dfile,'t2m',[1 1 expver hi],[inf inf 1 1])-278.5651;
    end
    temp=interp2(glat,glon,d1,pos(1),pos(2));
    d2=ncread(dfile,'sd',[1 1 expver hi],[inf inf 1 1]);
    sd=interp2(glat,glon,d2,pos(1),pos(2));
    d3=ncread(dfile,'sf',[1 1 expver hi],[inf inf 1 1]);
    sf=interp2(glat,glon,d3,pos(1),pos(2));
    data(hi,:)=[hzp(hi) temp sd sf];
    % Plot
    if plotmonths
        hg=geoshow(glat,glon,d3(:,:)*1000,'displaytype','surface');
        title(['\fontsize{20}{' datestr(hzp(hi),'mm/yyyy') '}']);
        drawnow
        delete(hg)
    end
end
%% Plot of data
% interpolated for given coordinate
figure('color',[1 1 1],'position',[100 100 600 1000])
subplot(3,1,1)
plot(data(:,1),data(:,2))
ylabel('°C');
datetick
title('mean air temperature in 2m');
subplot(3,1,2)
plot(data(:,1),data(:,3))
ylabel('m');
datetick
title('snow depth in m of water equivalent');
subplot(3,1,3)
plot(data(:,1),data(:,4)*1000)
ylabel('mm');
datetick
title('snow fall in mm of water equivalent');

File: ECMWF/ERA5_snow.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
ECMWF datafile access
 example for access of data in a netCDF file and interpolation of a single coordinate.
 t2m: air temperature in 2m height
 sd: snow depth expressed as a water layer covering the grid cell
 sf: snowfall expressed as a water layer
"""

import numpy as np
from netCDF4 import Dataset
import cftime
from scipy.interpolate import interp2d
import matplotlib.pyplot as plt

# Leipzig
pos = [51.3, 12.37]
# Datafile downloaded with ERA5_snow.py
dfile = 'ERA5_snow_monthly025.nc'
# Open file
fid = Dataset(dfile)
print(fid)
# coordinate grid
lon = np.array(fid['longitude'])
lat = np.array(fid['latitude'])
glat,glon =np.meshgrid(lat,lon)
# time
nctime = np.array(fid['time'])
nctimeunit = fid.variables['time'].units
zp=[]
zp.append(cftime.num2pydate(nctime, units=nctimeunit))
nds = len(zp[0])
# Data
at2m = np.array(fid['t2m'])
asd = np.array(fid['sd'])
asf = np.array(fid['sf'])
# iterpolation
t2m = np.zeros(nds)
sd = np.zeros(nds)
sf = np.zeros(nds)
expver=0
for hi in range(0, nds):
    if at2m[hi,expver,1,1]==-32767:
        expver=1
    ifun = interp2d(lon,lat, at2m[hi,expver,:,:],kind='linear')
    t2m[hi] = ifun(pos[1],pos[0])-278.5651
    ifun = interp2d(lon,lat, asd[hi,expver,:,:],kind='linear')
    sd[hi] = ifun(pos[1],pos[0])
    ifun = interp2d(lon,lat, asf[hi,expver,:,:],kind='linear')
    sf[hi] = ifun(pos[1],pos[0])
    
fig, ax = plt.subplots(3,1,figsize=(7,9))
plt.subplots_adjust(hspace=.0)
ax[0].plot_date(np.transpose(zp),t2m,'-')
ax[0].set_ylabel('Mean air temp at 2m / °C')
ax[0].grid()
ax[1].plot_date(np.transpose(zp),sd*1000,'-')
ax[1].set_ylabel('Snowdepth water equiv. /  mm')
ax[1].grid()
ax[2].plot_date(np.transpose(zp),sf*1000,'-')
ax[2].set_ylabel('Snowfall water equiv. /  mm')
ax[2].grid()

plt.show()